### PR TITLE
Swap upgrade: Allow users to filter exchanges and preview multiple quotes from different providers

### PR DESCRIPTION
**Pull request by justindg**
**Disclaimer**: Need @colourfreak 's input and final decision on everything related to UX.

---

Major changes:
1. Allow users to select Exchanges from the Settings screen. These can be pulled from the `/tools` endpoint, but I saved this to the local assets and modified it to provide the tool url/website. See `app/assets/swap_providers_list.json`.
306534239_469654408237434_7327828487150887135_n
306623672_5533693456712005_640825389844631434_n

2. Provide the users with a set of routes to choose from, based on the preferred Exchanges in Settings. In order to achieve this, we need to call [`/advanced/routes`](https://apidocs.li.fi/reference/post_advanced-routes), then use the selected Exchange key as a parameter to the `/quote` endpoint.

306706323_421049369934622_7034799854930041884_n

Minor changes:
1. Display provider/exchange website in the quote summary (clickable link).
307117477_2455068991300198_6934673362642555318_n

2. Display fees other than gas, if applicable. (Some exchanges have extra fees - can be found under `feeCosts`)
3. Refactored classes to provide more clarity/consistency

Issues fixed:
1. Fixed TokenSelector crashing on invalid input (decimal point)
2. Fixed wrong calculation of current price/rate
3. Fixed Gas Widget crash

### CODE REVIEW

**Comment by JamesSmartCell** [comment]
@justindg this is pretty epic! It's quite a big update I'll check it out and get back to you with some feedback

**Comment by JamesSmartCell** [comment]
If you bring it up to date can you do ```git rebase -i``` rather than ```git merge``` because otherwise it's really difficult to look at the new code.

File: app/src/main/java/com/alphawallet/app/ui/widget/entity/ProgressInfo.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package com.alphawallet.app.ui.widget.entity;
+
+public class ProgressInfo
+{
+    boolean shouldShow;
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/ui/widget/entity/ProgressInfo.java:5]: Can be `private`?

File: app/src/main/java/com/alphawallet/app/util/SwapUtils.java [hunk 2]
```diff
@@ -1,45 +1,69 @@
...
+import com.alphawallet.app.entity.lifi.FeeCost;
+import com.alphawallet.app.entity.lifi.GasCost;
 import com.alphawallet.app.entity.lifi.Quote;
 
 import java.math.BigDecimal;
 import java.util.ArrayList;
 
 public class SwapUtils
 {
+    public static final String CURRENT_PRICE_FORMAT = "1 %s ≈ %s %s";
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/util/SwapUtils.java:14]: Can be private.

File: app/src/main/java/com/alphawallet/app/viewmodel/SelectExchangesViewModel.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        Set<String> stringSet = new HashSet<>();
+        for (ToolDetails tool : selectedExchanges)
+        {
+            if (tool.isChecked)
+            {
+                stringSet.add(tool.key);
+                hasSelection = true;
+            }
+        }
+        if (hasSelection)
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/SelectExchangesViewModel.java:83]: Can check `!stringSet.isEmpty()` instead.

File: app/src/main/java/com/alphawallet/app/viewmodel/SwapViewModel.java [hunk 4]
```diff
@@ -360,6 +400,19 @@ public Web3Transaction buildWeb3Transaction(Quote quote)
         );
     }
 
+    public List<ToolDetails> getTools(Context context)
+    {
+        return new Gson().fromJson(Utils.loadJSONFromAsset(context, "swap_providers_list.json"),
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/SwapViewModel.java:405]: Duplicated.

**Review by seabornlee — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/alphawallet/app/viewmodel/SwapViewModel.java (hunk 4 above)
Comment (justindg) [reply app/src/main/java/com/alphawallet/app/viewmodel/SwapViewModel.java:405]: Did you mean to extract method into another class like `SwapUtils`?

Comment (seabornlee) [reply app/src/main/java/com/alphawallet/app/viewmodel/SwapViewModel.java:405]: Yes, maybe `SwapProviders.get()`.
And is the name `Tool` the same concept as SwapProvider/Exchange?
Should be better to have only one word for one concept.

**Comment by JamesSmartCell** [comment]
@justindg can you resolve merge conflict in the string file and do the translations? Then I think we can merge.

**Review by JamesSmartCell — APPROVED** [review]

**Review by seabornlee — APPROVED** [review]

**Pull request merged**